test/e2e/fpga/fpga.go

```diff
@@ -0,0 +1,143 @@
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+)
+
+const (
+	serviceAccountYaml   = "deployments/fpga_plugin/fpga_plugin_service_account.yaml"
+	pluginYaml           = "deployments/fpga_plugin/fpga_plugin.yaml"
+	webhookDeployScript  = "scripts/webhook-deploy.sh"
+	nlb0NodeResource     = "fpga.intel.com/af-f7df405cbd7acf7222f144b0b93acd18"
+	nlb0PodResource      = "fpga.intel.com/arria10.dcp1.2-nlb0"
+	nlb3PodResource      = "fpga.intel.com/arria10.dcp1.2-nlb3"
+	arria10dcp12Resource = "fpga.intel.com/region-69528db6eb31577a8c3668f9faa081f6"
+)
+
+func init() {
+	ginkgo.Describe("FPGA Plugin E2E tests", describe)
+}
+
+func describe() {
+	// Cleanup: delete plugin daemonset
+	// It's not deleted by ginkgo as it's in the kube-system namespace
+	ginkgo.AfterEach(func() {
+		ginkgo.By("removing FPGA plugin daemonset")
+		framework.RunKubectlOrDie("delete", "daemonset", "intel-fpga-plugin", "--namespace", "kube-system")
+	})
+
+	serviceAccountYamlPath, err := utils.LocateRepoFile(serviceAccountYaml)
+	if err != nil {
+		framework.Failf("unable to locate %q: %v", serviceAccountYaml, err)
+	}
+
+	pluginYamlPath, err := utils.LocateRepoFile(pluginYaml)
+	if err != nil {
+		framework.Failf("unable to locate %q: %v", pluginYaml, err)
+	}
+
+	webhookDeployScriptPath, err := utils.LocateRepoFile(webhookDeployScript)
+	if err != nil {
+		framework.Failf("unable to locate %q: %v", webhookDeployScript, err)
+	}
+
+	fmw := framework.NewDefaultFramework("fpgaplugin-e2e")
+
+	ginkgo.It("Run FPGA plugin tests", func() {
+		// Run region test case twice to ensure that device is reprogrammed at least once
+		runTestCase(fmw, pluginYamlPath, webhookDeployScriptPath, serviceAccountYamlPath, "region", "orchestrated", arria10dcp12Resource, nlb3PodResource, 3, 0)
+		runTestCase(fmw, pluginYamlPath, webhookDeployScriptPath, serviceAccountYamlPath, "region", "orchestrated", arria10dcp12Resource, nlb0PodResource, 0, 3)
+		// Run af test case
+		runTestCase(fmw, pluginYamlPath, webhookDeployScriptPath, serviceAccountYamlPath, "af", "preprogrammed", nlb0NodeResource, nlb0PodResource, 0, 3)
+	})
+}
+
+func runTestCase(fmw *framework.Framework, pluginYamlPath, webhookDeployScriptPath, serviceAccountYamlPath, pluginMode, webhookMode, nodeResource, podResource string, nlb1, nlb2 int16) {
+	ginkgo.By(fmt.Sprintf("deploying webhook in %s mode", webhookMode))
+	_, _, err := framework.RunCmd(webhookDeployScriptPath, "--mode", webhookMode, "--namespace", fmw.Namespace.Name)
+	framework.ExpectNoError(err)
+
+	ginkgo.By("creating service account for the plugin")
+	framework.RunKubectlOrDie("apply", "-f", serviceAccountYamlPath)
+
+	ginkgo.By(fmt.Sprintf("setting %s mode for the FPGA plugin", pluginMode))
+	annotation := fmt.Sprintf("fpga.intel.com/device-plugin-mode=%s", pluginMode)
+	framework.RunKubectlOrDie("annotate", "node", "--overwrite", "--all", annotation)
+
+	ginkgo.By("deploying FPGA plugin")
+	framework.RunKubectlOrDie("apply", "-f", pluginYamlPath)
```

Comment:
we had the kustomization overlays for these. did they not work?

Reply:
Running ```kubectl apply -k deployments/fpga_plugin/overlays/af/kustomization.yaml ``` from ginkgo fails to install webhook due to some certificate issues that I don't understand. I'm going to create separate issue for that.

Reply:
@mythi It could be that kustomisation fails due to hardcoded default namespace in the [webhook clusterrole.yaml](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/master/deployments/fpga_admissionwebhook/base/fpga-reader-clusterrole.yaml#L21)
Is it possible to have namespace parametrised somehow? Does kustomise support this?

Reply:
good question. with kustomize it's possible to patch the deployments but we don't know the namespace the test cases uses, or do we?

Reply:
namespaces can be changed this straightforward way:
```
diff --git a/deployments/fpga_plugin/overlays/region/kustomization.yaml b/deployments/fpga_plugin/overlays/region/kustomization.yaml
index 0b012fd..cdedf19 100644
--- a/deployments/fpga_plugin/overlays/region/kustomization.yaml
+++ b/deployments/fpga_plugin/overlays/region/kustomization.yaml
@@ -10,3 +10,5 @@ patchesJson6902:
       name: intel-fpga-plugin
     path:
       change-mode-region.yaml
+
+namespace: bla
```
```
> kustomize build overlays/region |grep namespace
  namespace: bla
      namespace: default
  namespace: bla
  namespace: kube-system
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
  namespace: bla
```

kustomise is clever enough to replace or add namespaces only where they needed, e.g. it didn't touch 'default' and 'kube-system' namespaces in ClusterRoleBindings.

However, this didn't make the whole solution to work. plugin and webhook pods start in specified namespace, but test workload pod fails to create with this error: ```Error from server: error when creating "test-fpga-region.yml": admission webhook "fpga.mutator.webhooks.intel.com" denied the request: Unknown region name: arria10```
It looks like kustomise doesn't configure CRD mappings correctly.

Reply:
Another problem with kustomize is that we'd need to generate kustomization.yaml with ginkgo namespace in it as kustomise doesn't do any variable substitutions except of container env, command and arguments.

Reply:
I think this is a bit out of scope of this PR, so I'm going to generate .yaml [the same way as webhook does.](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/master/scripts/webhook-deploy.sh#L107)

Reply:
`kubectl apply -k ... -n default` would also work so nothing needs to be generated

Reply:
it didn't work for me. Here is what I did:
1. patched the yaml:
```
--- a/deployments/fpga_plugin/base/intel-fpga-plugin-controller-serviceaccount.yaml
+++ b/deployments/fpga_plugin/base/intel-fpga-plugin-controller-serviceaccount.yaml
@@ -2,7 +2,7 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: intel-fpga-plugin-controller
-  namespace: kube-system
+  #namespace: kube-system
 ---
 kind: ClusterRole
 apiVersion: rbac.authorization.k8s.io/v1
@@ -21,7 +21,7 @@ metadata:
 subjects:
 - kind: ServiceAccount
   name: intel-fpga-plugin-controller
-  namespace: kube-system
+  #namespace: kube-system
 roleRef:
   kind: ClusterRole
   name: node-getter
```
2. > kubectl create -n default -k ./deployments/fpga_plugin/overlays/af
```
customresourcedefinition.apiextensions.k8s.io/acceleratorfunctions.fpga.intel.com created
customresourcedefinition.apiextensions.k8s.io/fpgaregions.fpga.intel.com created
mutatingwebhookconfiguration.admissionregistration.k8s.io/fpga-mutator-webhook-cfg created
serviceaccount/intel-fpga-plugin-controller created
clusterrole.rbac.authorization.k8s.io/fpga-reader created
clusterrole.rbac.authorization.k8s.io/node-getter created
clusterrolebinding.rbac.authorization.k8s.io/default-fpga-reader created
secret/intel-fpga-webhook-certs created
service/intel-fpga-webhook-svc created
deployment.apps/intel-fpga-webhook-deployment created
daemonset.apps/intel-fpga-plugin created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-compress created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-nlb3 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.1-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.1-nlb3 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.2-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.2-nlb3 created
acceleratorfunction.fpga.intel.com/d5005-nlb0 created
acceleratorfunction.fpga.intel.com/d5005-nlb3 created
fpgaregion.fpga.intel.com/arria10.dcp1.0 created
fpgaregion.fpga.intel.com/arria10.dcp1.1 created
fpgaregion.fpga.intel.com/arria10.dcp1.2 created
fpgaregion.fpga.intel.com/d5005 created
**The ClusterRoleBinding "get-nodes" is invalid: subjects[0].namespace: Required value**
```

Reply:
However, it worked this way just fine:
```
--- a/deployments/fpga_plugin/base/intel-fpga-plugin-controller-serviceaccount.yaml
+++ b/deployments/fpga_plugin/base/intel-fpga-plugin-controller-serviceaccount.yaml
@@ -2,7 +2,7 @@ apiVersion: v1
 kind: ServiceAccount
 metadata:
   name: intel-fpga-plugin-controller
-  namespace: kube-system
+  namespace: default
 ---
 kind: ClusterRole
 apiVersion: rbac.authorization.k8s.io/v1
@@ -21,7 +21,7 @@ metadata:
 subjects:
 - kind: ServiceAccount
   name: intel-fpga-plugin-controller
-  namespace: kube-system
+  namespace: default
 roleRef:
   kind: ClusterRole
   name: node-getter
```
```
> kubectl create -n default -k ./deployments/fpga_plugin/overlays/af
customresourcedefinition.apiextensions.k8s.io/acceleratorfunctions.fpga.intel.com created
customresourcedefinition.apiextensions.k8s.io/fpgaregions.fpga.intel.com created
mutatingwebhookconfiguration.admissionregistration.k8s.io/fpga-mutator-webhook-cfg created
serviceaccount/intel-fpga-plugin-controller created
clusterrole.rbac.authorization.k8s.io/fpga-reader created
clusterrole.rbac.authorization.k8s.io/node-getter created
clusterrolebinding.rbac.authorization.k8s.io/get-nodes created
clusterrolebinding.rbac.authorization.k8s.io/default-fpga-reader created
secret/intel-fpga-webhook-certs created
service/intel-fpga-webhook-svc created
deployment.apps/intel-fpga-webhook-deployment created
daemonset.apps/intel-fpga-plugin created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-compress created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.0-nlb3 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.1-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.1-nlb3 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.2-nlb0 created
acceleratorfunction.fpga.intel.com/arria10.dcp1.2-nlb3 created
acceleratorfunction.fpga.intel.com/d5005-nlb0 created
acceleratorfunction.fpga.intel.com/d5005-nlb3 created
fpgaregion.fpga.intel.com/arria10.dcp1.0 created
fpgaregion.fpga.intel.com/arria10.dcp1.1 created
fpgaregion.fpga.intel.com/arria10.dcp1.2 created
fpgaregion.fpga.intel.com/d5005 created
```

And plugin doesn't complain on start, i.e. it was able to get node annotation
```
> kubectl logs intel-fpga-plugin-t9kj8
FPGA device plugin (OPAE) started in af mode
Start server for af-d8424dc4a4a3c413f89e433683f9040b at: /var/lib/kubelet/device-plugins/fpga.intel.com-af-d8424dc4a4a3c413f89e433683f9040b.sock
Start server for af-f7df405cbd7acf7222f144b0b93acd18 at: /var/lib/kubelet/device-plugins/fpga.intel.com-af-f7df405cbd7acf7222f144b0b93acd18.sock
Device plugin for af-d8424dc4a4a3c413f89e433683f9040b registered
Device plugin for af-f7df405cbd7acf7222f144b0b93acd18 registered
```

Reply:
I'll probably leave it this way just to make e2e tests working and leave the rest to our  kustomise experts.

Reply:
I've added a bullet for it to #318